src/hardware/adc_internal.rs

```diff
@@ -55,67 +61,93 @@ impl AdcInternal {
         let mut adc3 = adc3.enable();
         adc3.set_resolution(adc::Resolution::SIXTEENBIT);
 
-        AdcInternal {
-            adc1,
-            adc3,
-            supply,
-            outu,
-            outi,
+        AdcInternal { adc1, adc3, pins }
+    }
+
+    pub fn read(&mut self, ch: AdcChannel) -> f32 {
+        match ch {
+            AdcChannel::OutputVoltage(ch) => self.read_output_voltage(ch),
+            AdcChannel::OutputCurrent(ch) => self.read_output_current(ch),
+            AdcChannel::OutputVref(ch) => self.read_output_vref(ch),
+            AdcChannel::Supply(ch) => self.read_supply(ch),
         }
     }
 
-    pub fn read_outu(&mut self, outu: OutChannel) -> u32 {
-        match outu {
-            OutChannel::Zero => self.adc1.read(&mut self.outu.0).unwrap(),
-            OutChannel::One => self.adc1.read(&mut self.outu.1).unwrap(),
-            OutChannel::Two => self.adc1.read(&mut self.outu.2).unwrap(),
-            OutChannel::Three => self.adc1.read(&mut self.outu.3).unwrap(),
+    pub fn read_output_voltage(&mut self, ch: OutputChannel) -> f32 {
+        let p = &mut self.pins.output_voltage;
+        let code: u32 = match ch {
+            OutputChannel::Zero => self.adc1.read(&mut p.0),
+            OutputChannel::One => self.adc1.read(&mut p.1),
+            OutputChannel::Two => self.adc1.read(&mut p.2),
+            OutputChannel::Three => self.adc1.read(&mut p.3),
         }
+        .unwrap();
+        const SCALE: f32 = V_REF * 20.0 / 5.0; // Differential voltage sense gain
+        const OFFSET: f32 = -VREF_TEC / V_REF; // Differential voltage sense offset
+        (code as f32 / self.adc1.max_sample() as f32 + OFFSET) * SCALE
     }
 
-    pub fn read_outi(&mut self, outi: OutChannel) -> u32 {
-        match outi {
-            OutChannel::Zero => self.adc1.read(&mut self.outi.0).unwrap(),
-            OutChannel::One => self.adc1.read(&mut self.outi.1).unwrap(),
-            OutChannel::Two => self.adc1.read(&mut self.outi.2).unwrap(),
-            OutChannel::Three => self.adc1.read(&mut self.outi.3).unwrap(),
+    pub fn read_output_current(&mut self, ch: OutputChannel) -> f32 {
+        let p = &mut self.pins.output_current;
+        let code: u32 = match ch {
+            OutputChannel::Zero => self.adc1.read(&mut p.0),
+            OutputChannel::One => self.adc1.read(&mut p.1),
+            OutputChannel::Two => self.adc1.read(&mut p.2),
+            OutputChannel::Three => self.adc1.read(&mut p.3),
         }
+        .unwrap();
+        const SCALE: f32 = V_REF / R_SENSE / 8.0; // MAX1968 ITEC scale
+        const OFFSET: f32 = -VREF_TEC / V_REF; // MAX1968 ITEC offset
+        (code as f32 / self.adc1.max_sample() as f32 + OFFSET) * SCALE
```

Comment:
Didn't double check this. @SingularitySurfer could you derive this independently based on the schematic and datasheet?

Reply:
I also looked at this and I think its correct.